core/src/main/java/org/lflang/generator/PortInstance.java

```diff
@@ -185,6 +198,50 @@ public List<RuntimeRange<PortInstance>> getDependsOnPorts() {
     return dependsOnPorts;
   }
 
+  /**
+   * For the given upstream port (the instance of a range returned by getDependsOnPorts()), return a
+   * list of connections from that port or null if there are no connections from that port.
+   *
+   * @param upstreamPort The upstream port.
+   */
+  public List<Connection> connectionsFrom(PortInstance upstreamPort) {
+    return upstreamConnections.get(upstreamPort);
+  }
+
+  /**
+   * For the given upstream port (the instance of a range returned by getDependsOnPorts()), return
+   * the minimum of the `after` delays on connections from that port. This will be TimeValue.ZERO if
+   * there are no `after` delays or if the `after` delays are zero, and it will be
+   * TimeValue.MAX_VALUE if there are no connections from the port.
+   *
+   * @param upstreamPort The upstream port.
+   */
+  public TimeValue minDelayFrom(PortInstance upstreamPort) {
+    TimeValue result = TimeValue.MAX_VALUE;
+    var connections = connectionsFrom(upstreamPort);
+    if (connections == null) return result;
+    for (var connection : connections) {
+      var delay = connection.getDelay();
+      if (delay != null) {
+        // The delay may reference a parameter defined in the parent of the connection,
+        // which may not be the same as the parent of the port, so resolve the delay using that
+        // parent.
+        var parent = upstreamPort.parent;
+        if (!parent.getDefinition().equals(connection.eContainer())) {
```

Comment:
This condition might be `true` _always_ because a port's parent reactor will not define an outside connection's parameter. But this can be a little subtle. I only realize this when seeing the diagram of `DelayToPhysicalAction.lf`.

Reply:
Actually, I think it can be either. `upstreamPort` may be an input, and then the `connection` will belong to the same parent as the port. But if `upstreamPort` is an output, then the connection belongs to the parent of the parent of the port. The problem is that we can't get the correct *instance* of the parent from the connection. The connection represents a syntactic object in the AST, but there can be many *instances* of this syntactic object. If we don't get the right instance, then we may not get the correct parameter values (because parameters can be overridden on a per-instance basis).

Reply:
I see. Yes, if `upstreamPort` is an input, then the connection will belong to the same parent.